File: swift/279_PerfectSquares.swift
class Solution {
    func numSquares(_ n: Int) -> Int {
        var squares = [Int]()
        var i = 0
        while i*i <= n {
            squares.append(i * i)
            if i*i == n { return 1 }
            i += 1
        }
        
        var count = n, head = squares.count - 1
        while head > 0 {
            for i in stride(from: head, to: 0, by: -1) {
                var currentCandidate = squares[head], checkingBackwardsHead = i, currentCount = 1
                while checkingBackwardsHead > 0 && currentCandidate < n && currentCount < count {
                    if currentCandidate + squares[checkingBackwardsHead] == n {
                        count = min(count, currentCount + 1)
                        break
                    } else if currentCandidate + squares[checkingBackwardsHead] > n {
                        checkingBackwardsHead -= 1
                    } else {
                        currentCount += 1
                        currentCandidate += squares[checkingBackwardsHead]
                    }
                }
            }
            head -= 1
        }
        return count
    }
}

File: java/279_PerfectSquares.java
class Solution {
    public int numSquares(int n) {
        List<Integer> squares = new ArrayList<Integer>();
        int size = 0;
        for (int i = 1; i*i <= n; ++i) {
            squares.add(i*i);
            if (squares.get(size++) == n) return 1;
        }
        
        int count = n;
        while (size > 1) {
            for (int i = size - 1; i > 0; --i) {
                int candidate = squares.get(size - 1), currCount = 1, head = i;
                while (candidate < n && currCount < count) {
                    if (candidate + squares.get(head) == n) {
                        count = Math.min(currCount + 1, count);
                        break;
                    } else if (candidate + squares.get(head) > n) {
                        --head;
                    } else {
                        currCount++;
                        candidate += squares.get(head);
                    }
                }
            }
            squares.remove(--size);
        }
        return count;
    }
}
